### PR TITLE
livecheck: update URL patterns

### PR DESCRIPTION
**Pull request by vladimyr**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

This updates `gnome`, `hackage`, `npm` & `pypi` livecheck strategies as discussed here: https://github.com/Homebrew/brew/pull/9510#discussion_r541326964

/cc @samford

### CODE REVIEW

File: Library/Homebrew/livecheck/strategy/hackage.rb [hunk 1]
```diff
@@ -18,7 +18,7 @@ module Strategy
       # @api public
       class Hackage
         # The `Regexp` used to determine if the strategy applies to the URL.
-        URL_MATCH_REGEX = /(?:downloads|hackage)\.haskell\.org/i.freeze
+        URL_MATCH_REGEX = %r{^https?://(?:downloads|hackage)\.haskell\.org(?:/[^/]+){3}}i.freeze
```
Comment (vladimyr) [inline Library/Homebrew/livecheck/strategy/hackage.rb:21]: Ensures that the path has exactly 3 parts because 3rd one is the filename used for package name extraction.
```rb
# Hackage URLs take one of the following formats:
#
# * `https://hackage.haskell.org/package/example-1.2.3/example-1.2.3.tar.gz`
# * `https://downloads.haskell.org/~ghc/8.10.1/ghc-8.10.1-src.tar.xz`
```

File: Library/Homebrew/livecheck/strategy/npm.rb [hunk 2]
```diff
@@ -20,7 +20,7 @@ class Npm
         NICE_NAME = "npm"
 
         # The `Regexp` used to determine if the strategy applies to the URL.
-        URL_MATCH_REGEX = /registry\.npmjs\.org/i.freeze
+        URL_MATCH_REGEX = %r{^https?://registry\.npmjs\.org(?:/[^/]+)?/[^/]+/-/}i.freeze
```
Comment (vladimyr) [inline Library/Homebrew/livecheck/strategy/npm.rb:23]:
```
%r{(?:/[^/]+)?/[^/]+/-/}i
```
covers both regular (`brew`) and scoped (`@brew/cli`) npm packages.

File: Library/Homebrew/livecheck/strategy/pypi.rb [hunk 3]
```diff
@@ -20,7 +20,7 @@ class Pypi
         NICE_NAME = "PyPI"
 
         # The `Regexp` used to determine if the strategy applies to the URL.
-        URL_MATCH_REGEX = /files\.pythonhosted\.org/i.freeze
+        URL_MATCH_REGEX = %r{^https?://files\.pythonhosted\.org/packages(?:/[^/]+){4}i}.freeze
```
Comment (vladimyr) [inline Library/Homebrew/livecheck/strategy/pypi.rb:23]: Ensures that the path has exactly 4 parts because 4th is the filename used for package name extraction.

```rb
`https://files.pythonhosted.org/packages/<hex>/<hex>/<long_hex>/example-1.2.3.tar.gz`
```

**Comment by samford** [comment]
I've exhausted the number of hours in this day, so I'll try to review/test this tomorrow to ensure it meets our needs and functions like we expect.

**Comment by vladimyr** [comment]
Take your time. I did this before reading your comment about promising to not touch other stuff until docs are ready. Now I feel guilty about pulling you into this right now. It can wait! 😉

**Review by samford — COMMENTED** [review]
I took some time to review this today and it's going in the right direction but I think we can take the idea further.

In a recent PR, @reitermarkus and I decided that it would be ideal if `URL_MATCH_REGEX` included the information for URL matching (in `#match?`) *and* extracting information from the URL (in `#find_versions`). This approach basically means that if the URL doesn't match the strategy, `#find_versions` wouldn't succeed in that case (and vice versa).

Since this PR relates to these ideas, I'm going to build on your work by pushing a commit that refactors related strategies (those other than `Bitbucket`, `Git`, `HeaderMatch`, `PageMatch`, and `Sparkle`) to follow the same pattern of `URL_MATCH_REGEX` reuse seen in #10068. These changes should resolve the mismatch between `URL_MATCH_REGEX` and the regex(es) used to extract information from the URL in `#find_versions`.

For the strategies that used more than one regex to extract information from the URL, I was able to simplify related logic by combining the regexes into `URL_MATCH_REGEX` using a `|` to allow for different scenarios. This notably simplifies `Gnu#find_versions` and `Sourceforge#find_version` while also making `#match?` more accurate.

Overall, I think the refactoring is a good improvement but I'm interested in feedback on these changes. I've done comparative before/after runs across homebrew/core and the only changes I saw were expected ones (e.g., `Sourceforge` no longer matches ~3 URLs that it incorrectly matched before due the looser regex).

I'm going to go through this again tomorrow with fresh eyes but I wanted to push my work so others can review it in the interim time.

**Review by samford — APPROVED** [review]
I went through this again today and addressed a few minor areas. Afterward, I did another comparative livecheck run (between `master` and this PR branch) and I didn't see any regressions.

I'm good with merging this once CI is working again and this passes but I didn't approve this earlier because I was hoping someone would review this to double-check my refactoring work.

File: Library/Homebrew/livecheck/strategy/pypi.rb [hunk 4]
```diff
@@ -19,8 +19,21 @@ module Strategy
...
+          .*? # The version string
+          (?<suffix>\.tar\.[a-z0-9]+|\.[a-z0-9]+)$ # Filename extension
+        /ix.freeze
+
         # The `Regexp` used to determine if the strategy applies to the URL.
-        URL_MATCH_REGEX = /files\.pythonhosted\.org/i.freeze
+        URL_MATCH_REGEX = %r{
+          ^https?://files\.pythonhosted\.org
+          /packages
+          (?:/[^/]+)+ # The hexadecimal paths before the filename
```
Comment (vladimyr) [inline Library/Homebrew/livecheck/strategy/pypi.rb:34]: Exactly three of them so maybe it is a good idea to be strict about it:
```suggestion
          (?:/[^/]+){3} # The hexadecimal paths before the filename
```

File: Library/Homebrew/livecheck/strategy/hackage.rb [hunk 5]
```diff
@@ -17,8 +17,19 @@ module Strategy
...
+        # also as part of extracting the package name from the URL basename.
+        PACKAGE_NAME_REGEX = /(?<package_name>.+?)-\d+/i.freeze
+
+        # A `Regexp` used to extract the package name from the URL basename.
+        FILENAME_REGEX = /^#{PACKAGE_NAME_REGEX.source.strip}/i.freeze
+
+        # A `Regexp` used in determining if the strategy applies to the URL.
+        URL_MATCH_REGEX = %r{
+          ^https?://(?:downloads|hackage)\.haskell\.org
+          /.+
```
Comment (vladimyr) [inline Library/Homebrew/livecheck/strategy/hackage.rb:30]:
```rb
# Hackage URLs take one of the following formats:
#
# * `https://hackage.haskell.org/package/example-1.2.3/example-1.2.3.tar.gz`
# * `https://downloads.haskell.org/~ghc/8.10.1/ghc-8.10.1-src.tar.xz`
```
There are always exactly two path segments before the filename:
```suggestion
          (?:/[^/]+){2}
```

**Review by vladimyr — COMMENTED** [review]
@samford Just so it is perfectly clear I feel bad suggesting these nitpicks because this is a truly heroic effort!

**Comment by vladimyr** [comment]
> In a recent PR, @reitermarkus and I decided that it would be ideal if `URL_MATCH_REGEX` included the information for URL matching (in `#match?`) _and_ extracting information from the URL (in `#find_versions`). This approach basically means that if the URL doesn't match the strategy, `#find_versions` wouldn't succeed in that case (and vice versa).

I mostly agree with this. IMHO swap _ideal_ with _practical_ and I'm fully on board. The fact that `#match?` uses regex with capturing groups is kinda weird. `#find_versions` applies the same regex again which makes me wonder do we really need both methods? :thinking: Is there an argument against merging `#match?` and `#find_versions` (to avoid invoking `regex#match/?` twice) into `#match_versions` with a bit complex return type?

> These changes should resolve the mismatch between `URL_MATCH_REGEX` and the regex(es) used to extract information from the URL in `#find_versions`.
> 
> For the strategies that used more than one regex to extract information from the URL, I was able to simplify related logic by combining the regexes into `URL_MATCH_REGEX` using a `|` to allow for different scenarios. This notably simplifies `Gnu#find_versions` and `Sourceforge#find_version` while also making `#match?` more accurate.

Unified, simplified, and more accurate - **beautiful**! :trophy: 

> I'm good with merging this but I didn't approve it earlier because I was hoping someone would review this to double-check all my refactoring work.

I would gladly eyeball it, even though I didn't feel confident touching more complex strategies initially but that is at least I can do to help you with this! Unfortunately, I'm far away from my usual work setup (discovering wonders of Github Codespaces on something that falls into glorified tablet bucket :tired_face:) and the most realistic forecast says it'll gonna stay that way till Jan 11th so I guess that makes me useless aside from glancing and potential nitpicks :shrug:

File: Library/Homebrew/livecheck/strategy/pypi.rb (hunk 4 above)
Comment (samford) [reply Library/Homebrew/livecheck/strategy/pypi.rb:34]: I'm not sure that there's a clear benefit to being explicit about this. We're not interested in the hexadecimal paths and it doesn't matter all that much how many there are. If it suddenly changed to any number other than three in the future (unlikely but possible), this would quietly fail to match those URLs until we manually update it.

This is just my own view but it seems like it would be better not to make more work for ourselves if we don't stand to benefit in some way from this being more explicit. We usually try to strike a balance between being too loose and too explicit, so I think it may be better to wait until this particular looseness creates a demonstrable problem for us.

File: Library/Homebrew/livecheck/strategy/hackage.rb (hunk 5 above)
Comment (samford) [reply Library/Homebrew/livecheck/strategy/hackage.rb:30]: Same as above. We're interested in the filename at the end of the path, so the number of directories doesn't matter all that much. The existing URL examples use two directories but this isn't necessarily guaranteed, so I think there's something to be said for being flexible in parts of the regex that aren't as important. I could understand if we were matching more URLs than we should be but that isn't the case yet.

In general, it's arguably easier to identify when a strategy is being applied to a URL it shouldn't match compared to identifying when a strategy isn't being applied to a URL it should match. Detecting the presence of something is often easier in livecheck than trying to detect the absence, so that's also part of my reasoning in terms of erring on the side of looseness here.

File: Library/Homebrew/livecheck/strategy/pypi.rb (hunk 4 above)
Comment (vladimyr) [reply Library/Homebrew/livecheck/strategy/pypi.rb:34]: > it doesn't matter all that much how many there are. If it suddenly changed to any number other than three in the future (unlikely but possible), this would quietly fail to match those URLs until we manually update it.
> 
> This is just my own view but it seems like it would be better not to make more work for ourselves if we don't stand to benefit in some way from this being more explicit.

What you just described I see as an important communication signal. It would be good to be explicitly notified once PyPI decides to change assumptions we are building on (which I do find highly unlikely and almost impossible - it would require pip refactoring) because they might be built-in elsewhere in the codebase too. So for me, the _benefit_ is exactly in detecting when things start silently rearranging under our feet :upside_down_face: 
Yes, you can say it doesn't matter because if you make it less strict no action is needed but that is a risky take. Sometimes less strict might also mean with potential _bug inside_ and figuring it out might turn into a nightmare because you wouldn't be aware that outside foundations changed in the first place. I guess it could be considered defensive coding and FWIW compared to laziness and adaptability it counts more on my personal scale. :upside_down_face:

File: Library/Homebrew/livecheck/strategy/hackage.rb (hunk 5 above)
Comment (vladimyr) [reply Library/Homebrew/livecheck/strategy/hackage.rb:30]: Yeah, we are successfully bikeshedding about regexes ATM but I find it truly useful within a good measure and with the right people.

> Detecting the presence of something is often easier in livecheck than trying to detect the absence, so that's also part of my reasoning in terms of erring on the side of looseness here.

That is true but it also means detecting false presence is easier too same as detecting false absence. My today's _hot take_ is that regexes aren't really about detecting presence/absence but matching (possibly fully qualified) form. It is borderline comparable to designing a keyhole pattern from an existing key, and it is always a good idea to be specific as possible to ensure only your (set of) key/s opens that lock. Except obviously, livecheck doesn't protect anyone's money safe but that's how I formed my opinion :upside_down_face:

File: Library/Homebrew/livecheck/strategy/pypi.rb (hunk 4 above)
Comment (samford) [reply Library/Homebrew/livecheck/strategy/pypi.rb:34]: I understand your view but I don't think the additional strictness is beneficial from the standpoint of livecheck. If other parts of `brew` make assumptions that are no longer true after a URL format change, then they're likely to loudly fail anyway. We shouldn't be making strategies more fragile for the sake of this kind of alternative purpose.

All that we're concerned about in this context is whether `#find_versions` successfully finds versions and the data is appropriate. If the URL format changes and the looser regex allows this to continue working without modifications, then we've saved ourselves some work. This would mean that the URL changes were in a part that doesn't impact the functionality of `#find_versions`. If the URL format changes and the strategy fails, then we'll know about it anyway (i.e., we don't need to make the regex more explicit to achieve this).

The goal of `URL_MATCH_REGEX` is to ensure that the URL being provided to `#find_versions` will work in that method. However, the regex is only as explicit as it needs to be to achieve this without introducing false positives. The number of hexadecimal directories doesn't impact the behavior of `Pypi#find_versions`, so we don't need to formalize it at this point.

If the looseness in this part of the regex ends up impacting the behavior of `#find_versions` in the future, then we'll see this issue in livecheck's output and we simply can address it in the strategy.

Comment (vladimyr) [reply Library/Homebrew/livecheck/strategy/pypi.rb:34]: > We shouldn't be making strategies more fragile for the sake of this kind of alternative purpose.

Agree.

>  This would mean that the URL changes were in a part that doesn't impact the functionality of `#find_versions`. If the URL format changes and the strategy fails, then we'll know about it anyway (i.e., we don't need to make the regex more explicit to achieve this).

Hmm, the latter is definitely true but I'm not universally sure about the former. I think it is still possible to make changes to the URL in a part that does in fact impact `#find_versions` but slips through `#match?`.

1. Now I need to prove it which requires coming up with some crazy clever pattern just for the sake of breaking it but I'm lazy :grin:  and 
2. We already agreed it doesn't matter at all because it would fail loudly anyway

Bottom line, from a practical standpoint you are right, and thank you for patiently addressing my philosophical arguments.

File: Library/Homebrew/livecheck/strategy/hackage.rb (hunk 5 above)
Comment (vladimyr) [reply Library/Homebrew/livecheck/strategy/hackage.rb:30]: So in the spirit of consistency and with acknowledging your arguments, here comes another proposal:
```suggestion
          (?:/[^/]+)+
```

This is a change for sake of change but I believe (yes it is subjective :shrug:) it better conveys the intent to the fellow programmer.

Comment (vladimyr) [reply Library/Homebrew/livecheck/strategy/hackage.rb:30]: Actually, following your line of thinking it would probably be better to invert it by changing the PyPI one to `/.+`. Whatever, but let's make it consistent if possible.

**Comment by samford** [comment]
> The fact that `#match?` uses regex with capturing groups is kinda weird. `#find_versions` applies the same regex again which makes me wonder do we really need both methods? 🤔 Is there an argument against merging `#match?` and `#find_versions` (to avoid invoking `regex#match/?` twice) into `#match_versions` with a bit complex return type?

`#match?` generally calls `Regexp#match?`, which checks whether the whole regex matches the URL string. It doesn't make a difference if the regex contains capture groups or not, as `Regexp#match?` doesn't do anything with them. On the other hand, `Regexp#match` (as seen in `#find_versions`) does use the capture groups found in the regex. These two methods serve different purposes and, as you may expect, `Regexp#match?` does less work than `Regexp#match` (and is 2-3x faster as a result).

Each strategy's `#match?` method is used to determine which strategies apply to a given URL. When testing a given URL, we execute every strategy's `#match?` method as part of this process. We don't execute a strategy's `#find_versions` method until we've identified all the matching strategies and selected the highest priority one. If no strategies match, then we don't execute `#find_versions` for any strategies for that URL.

The short version is no, we shouldn't combine `#match?` and `#find_versions` and, unless I'm overlooking something, doing so would result in us doing more work rather than less. I'll try to explain in more detail tomorrow but it's very late over here and I need to sleep.

I'm labeling this "do not merge" for the moment, so I can address the feedback in the morning.

File: Library/Homebrew/livecheck/strategy/hackage.rb (hunk 5 above)
Comment (samford) [reply Library/Homebrew/livecheck/strategy/hackage.rb]: They both match the same text in this context but `(?:/[^/]+)+` is clear about what type of text it's working with, so I'm good with the replacement in `Hackage`. I incorporated it in the latest push.

**Comment by samford** [comment]
I'm working on some other changes that touch some of this code, so I'm going to go ahead and merge this. Thanks, @vladimyr!
